### PR TITLE
Setting for turning off filtering based on path

### PR DESCRIPTION
**Pull request by gnidorah**
#377
Thank you!

### CODE REVIEW

File: src/xpiks-qt/Helpers/filterhelpers.cpp [hunk 1]
```diff
@@ -194,10 +194,14 @@ namespace Helpers {
         return hasMatch;
     }
 
-    bool containsPartsSearch(const QString &mainSearchTerm, Models::ArtworkMetadata *metadata, bool searchUsingAnd) {
+    bool containsPartsSearch(const QString &mainSearchTerm, Models::ArtworkMetadata *metadata, bool searchUsingAnd, bool searchByFilepath) {
         // default search is not case sensitive
         Common::SearchFlags searchFlags = searchUsingAnd ? Common::SearchFlags::AllTermsEverything :
                                            Common::SearchFlags::AnyTermsEverything;
+        if (!searchByFilepath) {
+          Common::UnsetFlag(searchFlags, Common::SearchFlags::Filepath);
```
Comment (Ribtoks) [inline src/xpiks-qt/Helpers/filterhelpers.cpp]: There's a whitespace misalignment here - some spaces are missing.

File: src/xpiks-qt/Dialogs/SettingsWindow.qml [hunk 2]
```diff
@@ -301,6 +301,21 @@ ApplicationWindow {
                             }
                         }
 
+                        StyledCheckbox {
+                            id: searchByFilepathCheckbox
+                            text: i18.n + qsTr("Search by path")
+                            onCheckedChanged: {
+                                settingsModel.searchByFilepath = checked
```
Comment (Ribtoks) [inline src/xpiks-qt/Dialogs/SettingsWindow.qml]: What if there were some artworks in the main grid already filtered? After changing this setting they should be updated (by calling `invalidateFilter()` of FilteredArtworksModel). I propose to create a signal in SettingsModel smth like `settingsSaved()` or `settingsUpdated()` and slot in FilteredArtworksModel smth like `onSettingsUpdated()` which will `invalidateFilter()` as well as update Filter Flags - pls see below on this.

File: src/xpiks-qt/Models/filteredartitemsproxymodel.cpp [hunk 3]
```diff
@@ -639,8 +639,9 @@ namespace Models {
         if (metadata != NULL) {
             SettingsModel *settingsModel = m_CommandManager->getSettingsModel();
             bool searchUsingAnd = settingsModel->getSearchUsingAnd();
+            bool searchByFilepath = settingsModel->getSearchByFilepath();
 
-            hasMatch = Helpers::containsPartsSearch(m_SearchTerm, metadata, searchUsingAnd);
+            hasMatch = Helpers::containsPartsSearch(m_SearchTerm, metadata, searchUsingAnd, searchByFilepath);
```
Comment (Ribtoks) [inline src/xpiks-qt/Models/filteredartitemsproxymodel.cpp]: This was fine with 1 boolean, but now it grows and requires a lot of changes (look at your changes in tests). For this to be usable in future (maybe more booleans?), I propose to remove this method `Helpers::containsPartsSearch()` and use only `Helpers::hasSearchMatch()` directly. 

In order to do so, you need:

- add SearchFlags filed to _FilteredArtItemsProxyModel_ (along with m_SearchTerm)
- in the `onSettingsUpdated()` and `setSearchTerm()` you will need to reconstruct it (set/unset different flags)
- in the `FilteredArtItemsProxyModel::filterAcceptsRow()` use directly the flags and `Helpers::hasSearchMatch()` - there will be performance benefits over current setup
- fix tests (in order not to create flags all of the times - create 2 helper methods of same `ArtworkFilterTests` class which will return needed combinations of flags and use the methods directly in the `hasSearchMatch()` call (with which you will need to replace current `Helpers::containsPartsSearch`)

File: src/xpiks-qt/Models/filteredartitemsproxymodel.cpp [hunk 4]
```diff
@@ -54,6 +55,13 @@ namespace Models {
             emit searchTermChanged(value);
         }
 
+        SettingsModel *settingsModel = m_CommandManager->getSettingsModel();
+        bool searchUsingAnd = settingsModel->getSearchUsingAnd();
+        bool searchByFilepath = settingsModel->getSearchByFilepath();
+        // default search is not case sensitive
+        Common::ApplyFlag(m_SearchFlags, searchUsingAnd, Common::SearchFlags::AllTerms);
```
Comment (Ribtoks) [inline src/xpiks-qt/Models/filteredartitemsproxymodel.cpp]: This is not entirely correct way to do it. The correct one was in the removed `containsParts()`

    Common::SearchFlags searchFlags = searchUsingAnd ? Common::SearchFlags::AllTermsEverything :
                                        Common::SearchFlags::AnyTermsEverything;

File: src/xpiks-qt/Models/filteredartitemsproxymodel.cpp [hunk 5]
```diff
@@ -466,6 +474,15 @@ namespace Models {
         }
     }
 
+    void FilteredArtItemsProxyModel::onSettingsUpdated() {
+        LOG_DEBUG << "#";
+        SettingsModel *settingsModel = m_CommandManager->getSettingsModel();
```
Comment (Ribtoks) [inline src/xpiks-qt/Models/filteredartitemsproxymodel.cpp]: This is a duplicating code with the `setSearchTerm()`. A better way to do it is to create a method `updateSearchFlags()` or smth like that, put this code in there and then just use it in `onSettingsUpdated()` and `setSearchTerm()`

File: src/xpiks-qt/Dialogs/SettingsWindow.qml [hunk 6]
```diff
@@ -301,6 +301,22 @@ ApplicationWindow {
                             }
                         }
 
+                        StyledCheckbox {
+                            id: searchByFilepathCheckbox
+                            text: i18.n + qsTr("Search by path")
+                            onCheckedChanged: {
+                                settingsModel.searchByFilepath = checked
+                                settingsModel.updateSettings()
```
Comment (Ribtoks) [inline src/xpiks-qt/Dialogs/SettingsWindow.qml]: This is not a correct place to use this. What if the user just checked/unchecked the checkbox? See my comment in the settingsmodel.cpp file

File: src/xpiks-qt/Models/settingsmodel.h [hunk 7]
```diff
@@ -81,6 +82,7 @@ namespace Models {
     public:
         Q_INVOKABLE void resetAllValues();
         Q_INVOKABLE void saveAllValues();
+        Q_INVOKABLE void updateSettings() { emit settingsUpdated(); }
```
Comment (Ribtoks) [inline src/xpiks-qt/Models/settingsmodel.h]: It doesn't seem to me that this should be a separate method. Rather the signal should be emited the `saveAllValues()`

**Comment by gnidorah** [comment]
@Ribtoks Done. Sorry for making wrong assumptions!

File: src/xpiks-tests/xpiks-tests-core/artworkfilter_tests.h [hunk 8]
```diff
@@ -3,15 +3,25 @@
 
 #include <QObject>
 #include <QtTest/QtTest>
+#include "../../xpiks-qt/Common/flags.h"
 
 class ArtworkFilterTests : public QObject
 {
     Q_OBJECT
+private:
+    Common::SearchFlags flagsAnyFilepath() { return Common::SearchFlags::AnyTermsEverything; }
```
Comment (Ribtoks) [inline src/xpiks-tests/xpiks-tests-core/artworkfilter_tests.h]: Can you please rename `flagsAnyFilepath()` to `flagsAnyTermWithFilepath()` and `flagsAnd()` to `flagsAllTermsWithoutFilepath()`

**Comment by Ribtoks** [comment]
@gnidorah All good, but you removed initialization code for `m_SearchFlags` in the Filtered Model's constructor. Have you tried launching Xpiks with your changes and searching by filepath? Please initialize it to `Common::SearchFlags::AnyTermsEverything` in the constructor.

**Pull request merged**